### PR TITLE
Update cache from raft StateMachine

### PR DESCRIPTION
**Pull request by Linary**
Change-Id: I7b57ed687a9f12c00a6561f0c1590c27247d47c1

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 1]
```diff
@@ -199,7 +138,7 @@ public Object submitAndWait(StoreCommand command, StoreClosure closure) {
         }
     }
 
-    private void waitLeaderElected() {
+    public void waitLeaderElected() {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:141]: set protected

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java [hunk 2]
```diff
@@ -95,6 +105,82 @@ public void addNode(String group, BackendStore store) {
...
+        NodeOptions nodeOptions = new NodeOptions();
+        int electionTimeout = config.get(CoreOptions.RAFT_ELECTION_TIMEOUT);
+        nodeOptions.setElectionTimeoutMs(electionTimeout);
+        nodeOptions.setDisableCli(false);
+
+        int snapshotInterval = config.get(CoreOptions.RAFT_SNAPSHOT_INTERVAL);
+        nodeOptions.setSnapshotIntervalSecs(snapshotInterval);
+
+        Configuration initConf = new Configuration();
+        String initConfStr = config.get(CoreOptions.RAFT_GROUP_PEERS);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java:122]: rename

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java [hunk 3]
```diff
@@ -92,6 +97,24 @@ private void registerCommands() {
         });
     }
 
+    private void updateCacheIfNeeded(BackendMutation mutation) {
+        // Only follower need to update cache from store to tx
+        if (this.isLeader()) {
+            return;
+        }
+        for (HugeType type : mutation.types()) {
+            if (!type.isGraph() && type.isSchema()) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java:106]: !type.isSchema()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java [hunk 4]
```diff
@@ -92,6 +97,24 @@ private void registerCommands() {
...
+                continue;
+            }
+            for (java.util.Iterator<BackendAction> it = mutation.mutation(type);
+                 it.hasNext();) {
+                BackendAction item = it.next();
+                BackendEntry entry = item.entry();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java:112]: remove the define

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedGraphTransaction.java [hunk 5]
```diff
@@ -135,14 +135,16 @@ private void listenChanges() {
...
             LOG.debug("Graph {} received graph cache event: {}",
                       this.graph(), event);
-            event.checkArgs(String.class, Id.class);
+            event.checkArgs(String.class, HugeType.class, Id.class);
             Object[] args = event.args();
             if ("invalid".equals(args[0])) {
-                Id id = (Id) args[1];
-                if (this.verticesCache.get(id) != null) {
+                HugeType type = (HugeType) args[1];
+                Id id = (Id) args[2];
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedGraphTransaction.java:142]: clear edgesCache first

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedGraphTransaction.java [hunk 6]
```diff
@@ -135,14 +135,16 @@ private void listenChanges() {
...
+                if (type.isVertex()) {
                     // Invalidate vertex cache
                     this.verticesCache.invalidate(id);
-                } else if (this.edgesCache.get(id) != null) {
+                } else if (type.isEdge()) {
+                    id = new QueryId(new IdQuery(type, id));
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedGraphTransaction.java:147]: QueryId is just for query, id of edge-update don't need to wrap QueryId

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeLabelCoreTest.java [hunk 7]
```diff
@@ -1215,7 +1215,7 @@ public void testListEdgeLabels() {
         Assert.assertTrue(edgeLabels.contains(write));
 
         // clear cache
-        params().schemaEventHub().call(Events.CACHE, "clear", null);
+        params().schemaEventHub().call(Events.CACHE, "clear", null, null);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeLabelCoreTest.java:1218]: add get

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexLabelCoreTest.java [hunk 8]
```diff
@@ -1097,7 +1097,7 @@ public void testListVertexLabels() {
         Assert.assertTrue(vertexLabels.contains(book));
 
         // clear cache
-        params().schemaEventHub().call(Events.CACHE, "clear", null);
+        params().schemaEventHub().call(Events.CACHE, "clear", null, null);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexLabelCoreTest.java:1100]: ditto

Comment (Linary) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexLabelCoreTest.java:1100]: use call instead of notify

**Pull request merged**